### PR TITLE
Add utility to send telemetry

### PR DESCRIPTION
**Pull request by Wauplin**
Related to [internal discussion](https://huggingface.slack.com/archives/C02Q9UFS8AW/p1676391976224329) related to Gradio telemetry. Idea is to centralize telemetry across HF libraries in a single helper. Gradio could be a first step (cc @abidlabs @freddyaboulton @dawoodkhan82) , but it can also replace implementation from [`transformers`](https://github.com/huggingface/transformers/blob/main/src/transformers/utils/hub.py#LL125C87-L125C95) and `diffusers` (cc @sgugger @patrickvonplaten, even though this is not a priority).

Telemetry is sent to _https://huggingface.co/api/telemetry/whatever/path_ using a HEAD request. Headers contain a user_agent with library name/version + some stuff (hf_hub version, python version, pytorch/tf/...,). Additional data can be also provided in the user_agent, either with a string or with a dictionary. Entries are de-duplicated from the user_agent before being sent to the Hub.

Since telemetry is a bit sensible, it will be at the discretion of each library to decide when/where to put telemetry. There is always an opt-out option for the users to disable it globally with `HF_HUB_DISABLE_TELEMETRY` (or `HF_HUB_OFFLINE` which has [other side effects](https://huggingface.co/docs/huggingface_hub/v0.12.0/en/package_reference/environment_variables#hfhuboffline)). `DISABLE_TELEMETRY` is also supported as legacy for `transformers`. (We can also add helpers like `disable_telemetry`/`enable_telemetry` to programmatically enable/disable it we you think it's useful).

**Example:**
```py
>>> from huggingface_hub.utils import send_telemetry

# Send telemetry without library information
>>> send_telemetry("ping")

# Send telemetry to subtopic with library information
>>> send_telemetry("gradio/local_link", library_name="gradio", library_version="3.22.1")

# Send telemetry with additional data
>>> send_telemetry(
...     topic="examples",
...     library_name="transformers",
...     library_version="4.26.0",
...     user_agent={"pipeline": "text_classification", "framework": "flax"},
... )
```

**TODO:**
- [x] `send_telemetry`
- [x] unit tests
- [x] docs in package reference section
- [x] docs in environment variables sections 
- [x] (optional) run telemetry in a separate thread to avoid blocking the main thread in case of slow connections

cc @julien-c @LysandreJik who brought up the topic :)

### CODE REVIEW

**Review by julien-c — APPROVED** [review]
really nice implem!

File: src/huggingface_hub/utils/_telemetry.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    ...     user_agent={"pipeline": "text_classification", "framework": "flax"},
+    ... )
+    ```
+    """
+    if constants.HF_HUB_OFFLINE or constants.HF_HUB_DISABLE_TELEMETRY:
+        return
+
+    path = "/".join(quote(part) for part in topic.split("/") if len(part) > 0)
+    try:
+        r = requests.head(
```
[the remark this answers is not in the document]
Comment (julien-c) [reply src/huggingface_hub/utils/_telemetry.py:63]: yep, sounds good (the server doesn't return a payload, just a `200 OK`, but i understand the argument)

**Review by sgugger — APPROVED** [review]
Looks very nice, thanks!

**Comment by dawoodkhan82** [comment]
@Wauplin any update on when we can expect this to be merged?

**Comment by Wauplin** [comment]
@dawoodkhan82 It's tested and reviewed so let's merge it now :)

But I guess your question is more, "when will it be released"? For this question I'm not sure yet. Except if it's urgent I don't think I'll make a new release of `huggingface_hub` in the next 2 weeks. Would that be ok for you. Open to discussion of course :)

**Comment by dawoodkhan82** [comment]
@Wauplin Not urgent, but would be nice if it was released sooner. It would help with fully migrating away from segment sooner (which bills us monthly).

**Comment by julien-c** [comment]
@dawoodkhan82 but are you going to pin the latest huggingface_hub in gradio then? (just curious)

**Comment by dawoodkhan82** [comment]
@julien-c Yeah that's the plan. I think we should also collect analytics using both segment + `huggingface_hub` for the first 2 weeks or so to make sure the transition is smooth + there's no discrepancies. Shouldn't be an issue right? @abidlabs

**Pull request merged**